src/Workspaces/Core/Portable/Workspace/Solution/SourceGeneratedDocumentState.cs

```diff
@@ -25,11 +27,39 @@ internal sealed class SourceGeneratedDocumentState : DocumentState
     /// </summary>
     public SourceText SourceText { get; }
 
+    /// <summary>
+    /// Checksum of <see cref="SourceText"/> when it was <em>originally</em> created.  This is subtly, but importantly
+    /// different from the checksum acquired from <see cref="SourceText.GetChecksum"/>.  Specifically, the original
+    /// source text may have been created from a <see cref="System.IO.Stream"/> in a lossy fashion (for example,
+    /// removing BOM marks and the like) on the OOP side. As such, its checksum might not be reconstructible from the
+    /// actual text and hash algorithm that were used to create the SourceText on the host side.  To ensure both the
+    /// host and OOP are in agreement about the true content checksum, we store this separately.
+    /// </summary>
+    public Checksum GetOriginalSourceTextChecksum()
+        => _lazyTextChecksum.Value;
+
     public static SourceGeneratedDocumentState Create(
         SourceGeneratedDocumentIdentity documentIdentity,
         SourceText generatedSourceText,
         ParseOptions parseOptions,
-        LanguageServices languageServices)
+        LanguageServices languageServices,
+        Checksum? originalSourceTextChecksum)
+    {
+        // If the caller explicitly provided us with the checksum for the source text, then we always defer to that.
+        // This happens on the host side, when we are give the data computed by the OOP side.
```

Comment:
>give [](http://example.com/codeflow?start=54&length=4)

nit: given

Reply:
lets not bump up another commit on this given the time sensitivity of the 17.9P1 build respin needed